sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -475,6 +475,8 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
       case If(TrueLiteral, trueValue, _) => trueValue
       case If(FalseLiteral, _, falseValue) => falseValue
       case If(Literal(null, _), _, falseValue) => falseValue
+      case If(_, TrueLiteral, TrueLiteral) => TrueLiteral
+      case If(_, FalseLiteral, FalseLiteral) => FalseLiteral
       case If(cond, TrueLiteral, FalseLiteral) => cond
       case If(cond, FalseLiteral, TrueLiteral) => Not(cond)
       case If(cond, trueValue, falseValue)
```

Comment:
@wangyum, doesn't this cover the case above by this case match?

Reply:
This is only for deterministic expressions.

Reply:
The conditions you added here also should only for deterministic ones then because `cond` will not be executed, see https://github.com/apache/spark/pull/21848#issuecomment-407166299 as an example.

Reply:
BTW, the tests you added at https://github.com/apache/spark/pull/30898/files#diff-c18cf81d53bb8508802317a785a74f2dadb0487d88e0164efb36f89ce64973e3R210-R213 will pass because of the case match here.